src/with-props.js

```diff
@@ -0,0 +1,176 @@
+import { withRaw } from './with-raw';
+import {
+  debounce,
+  keys,
+  uniqueId
+} from './util';
+import {
+  _updateDebounced,
+  defineProps,
+  normPropDefs,
+  syncAttributeToProperty
+} from './util/with-props';
+
+const _connected = uniqueId();
+const _observedAttributes = uniqueId();
+const _prevProps = uniqueId();
+const _props = uniqueId();
+const _updateCallback = uniqueId();
+const _updating = uniqueId();
+
+export function withProps (Base = withRaw()) {
+  return class extends Base {
+    static get observedAttributes () {
+      const props = normPropDefs(this);
+      return keys(props)
+        .map(k => props[k].attribute)
+        .filter(Boolean)
+        .map(a => a.source)
+        .concat(this[_observedAttributes] || []);
+    }
+
+    static set observedAttributes (value) {
+      this[_observedAttributes] = value;
+    }
+
+    static get props () {
+      return { ...super.props, ...this[_props] };
+    }
+
+    static set props (value) {
+      this[_props] = value;
+    }
+
+    constructor () {
+      super();
+      const { constructor } = this;
```

Comment:
so correct me if I understood this correctly ( I'm asking because of setting "default/initial" values of property via Class property initialisers )

if I have following component:

```js
class MyCmp extends Component {
  static get props(){
     return {
        name: prop.string,
        year: prop.number,
     }
  }

  name = 'Rodney Mullen';
  year = 1998;

  renderCallback({name,age}){
    return (
      <p>First DarkSlide was done by <b>{name}</b> in <b>{age}</b></p>
    )
  }
  
}
```

At first getter/setter will be created from props on `this`
-> defineProps(constructor);
after that `name = 'Rodney Mullen';` ->transpiledTo-> `this.name = 'Rodney Mullen';` will properly invoke defined prop setter (`name`) and within it it will store the value to internal `_value` 

-> 
```
 set (val) {
 +        this[_value] = coerce(val);
 +        syncPropertyToAttribute(this, target, serialize, val);
 +        this[_updateDebounced]();
 +      }
```